hathor/manager.py

```diff
@@ -633,59 +719,99 @@ def propagate_tx(self, tx: BaseTransaction, fails_silently: bool = True) -> bool
-                    raise InvalidNewTransaction('Transaction already exists {}'.format(tx.hash_hex))
-                self.log.debug('on_new_tx(): Transaction already exists', tx=tx.hash_hex)
-                return False
+        if self.tx_storage.transaction_exists(tx.hash):
+            if not fails_silently:
+                raise InvalidNewTransaction('Transaction already exists {}'.format(tx.hash_hex))
+            self.log.warn('on_new_tx(): Transaction already exists', tx=tx.hash_hex)
+            return False
 
-        if self.state != self.NodeState.INITIALIZING or self._full_verification:
+        if tx.timestamp - self.reactor.seconds() > settings.MAX_FUTURE_TIMESTAMP_ALLOWED:
+            if not fails_silently:
+                raise InvalidNewTransaction('Ignoring transaction in the future {} (timestamp={})'.format(
+                    tx.hash_hex, tx.timestamp))
+            self.log.warn('on_new_tx(): Ignoring transaction in the future', tx=tx.hash_hex,
+                          future_timestamp=tx.timestamp)
+            return False
+
+        tx.storage = self.tx_storage
+
+        try:
+            metadata = tx.get_metadata()
+        except TransactionDoesNotExist:
+            if not fails_silently:
+                raise InvalidNewTransaction('missing parent')
+            self.log.warn('on_new_tx(): missing parent', tx=tx.hash_hex)
+            return False
+
+        if metadata.validation.is_invalid():
+            if not fails_silently:
+                raise InvalidNewTransaction('previously marked as invalid')
+            self.log.warn('on_new_tx(): previously marked as invalid', tx=tx.hash_hex)
+            return False
+
+        if not partial or (metadata.validation.is_fully_connected() or tx.can_validate_full()):
+            if isinstance(tx, Transaction) and self.tx_storage.is_tx_needed(tx.hash):
+                tx._height_cache = self.tx_storage.needed_index_height(tx.hash)
+
+            if not metadata.validation.is_valid():
+                try:
+                    tx.validate_full()
+                except HathorError as e:
+                    if not fails_silently:
+                        raise InvalidNewTransaction('full validation failed') from e
+                    self.log.warn('on_new_tx(): full validation failed', tx=tx.hash_hex, exc_info=True)
+                    return False
+
+            # The method below adds the tx as a child of the parents
+            # This needs to be called right before the save because we were adding the children
+            # in the tx parents even if the tx was invalid (failing the verifications above)
+            # then I would have a children that was not in the storage
+            tx.update_initial_metadata()
+            self.tx_storage.save_transaction(tx, add_to_indexes=True)
             try:
-                assert self.validate_new_tx(tx, skip_block_weight_verification=skip_block_weight_verification) is True
```

Comment:
I can't find where `self.validate_new_tx()` is being called.

Reply:
It isn't anymore. I think it can be removed.

Reply:
Should we remove this method?

Reply:
Removed.